src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Microbenchmarks/LanguageServer/RazorDiagnosticsBenchmark.cs

```diff
@@ -0,0 +1,227 @@
+using Microsoft.AspNetCore.Razor.LanguageServer;
+using Microsoft.AspNetCore.Razor.LanguageServer.Diagnostics;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.Microbenchmarks.LanguageServer;
+
+public class RazorDiagnosticsBenchmark : RazorLanguageServerBenchmarkBase
+{
+    private string? _filePath;
+    private Uri? DocumentUri { get; set; }
+    private DocumentPullDiagnosticsEndpoint? DocumentPullDiagnosticsEndpoint { get; set; }
+    private IDocumentSnapshot? DocumentSnapshot { get; set; }
+    private SourceText? DocumentText { get; set; }
+    private RazorRequestContext RazorRequestContext { get; set; }
+
+    public enum FileTypes
+    {
+        Small,
+        Large
+    }
+
+    [ParamsAllValues]
+    public FileTypes FileType { get; set; }
+
+    [GlobalSetup]
+    public async Task SetupAsync()
+    {
+        var languageServer = RazorLanguageServer.GetInnerLanguageServerForTesting();
+
+        DocumentPullDiagnosticsEndpoint = new DocumentPullDiagnosticsEndpoint(
+            languageServerFeatureOptions: languageServer.GetRequiredService<LanguageServerFeatureOptions>(),
+            translateDiagnosticsService: languageServer.GetRequiredService<RazorTranslateDiagnosticsService>(),
+            languageServer: languageServer.GetRequiredService<ClientNotifierServiceBase>());
+        var projectRoot = Path.Combine(RepoRoot, "src", "Razor", "test", "testapps", "ComponentApp");
+        var projectFilePath = Path.Combine(projectRoot, "ComponentApp.csproj");
+        _filePath = Path.Combine(projectRoot, "Components", "Pages", $"Generated.razor");
+
+        var content = GetFileContents(this.FileType);
+        File.WriteAllText(_filePath, content);
+
+        var targetPath = "/Components/Pages/Generated.razor";
+
+        DocumentUri = new Uri(_filePath);
+        DocumentSnapshot = GetDocumentSnapshot(projectFilePath, _filePath, targetPath);
+        DocumentText = await DocumentSnapshot.GetTextAsync();
+        var documentContext = new VersionedDocumentContext(DocumentUri, DocumentSnapshot, 1);
+
+        RazorRequestContext = new RazorRequestContext(documentContext, Logger, languageServer.GetLspServices());
+
+        // Run once in setup to verify setup is correct
+        var request = new VSInternalDocumentDiagnosticsParams
+        {
+            TextDocument = new TextDocumentIdentifier
+            {
+                Uri = DocumentUri!
+            },
+        };
+
+        var diagnostics = await DocumentPullDiagnosticsEndpoint!.HandleRequestAsync(request, RazorRequestContext, CancellationToken.None);
+
+        if (!diagnostics!.ElementAtOrDefault(0)!.Diagnostics!.ElementAtOrDefault(0)!.Message.Contains("CallOnMe"))
```

Comment:
fwiw, I imagine life would be easier with nullable disabled in benchmark code. It relies on something the compiler couldn't know about. If it's wrong... the benchmark was already wrong anyways.